lib/dot.js

```diff
@@ -1,166 +1,123 @@
-    // This allows swapping in different dot positioning strategies.
-    this.placement = placement
+    
+  // grid is the grid of opportunity densities to visualize as map tiles (class from Browsochrones). 
+  // color is the color of the dots to draw.
+  initialize (grid, color = '#49a0d7') {
     this.color = new Color(color)
-    // The grid of opportunity densities to visualize as map tiles.
     this.grid = grid
   }
 
   // Given a web Mercator x, y, and zoom values for a single tile, as well as an associated canvas object,
   // visualize the opportunities falling within that tile as dots on the canvas.
-  drawTile (canvas, { x, y }, zoom) {
+  drawTile (canvas, mercTileCoord, zoom) {
     const ctx = canvas.getContext('2d')
     const imageData = ctx.createImageData(canvas.width, canvas.height)
 
     // Convert web Mercator tile position to pixels relative to the left and the top of the world at its zoom level.
-    x *= 256
-    y *= 256
-
-    const transparency = TRANSPARENCY_BY_ZOOM[zoom]
-    const rgba = [...this.color.rgbArray(), transparency]
+    const mercPixelX = mercTileCoord.x * 256
+    const mercPixelY = mercTileCoord.y * 256
 
     // Compute the divisor that will convert pixel (or tile) coordinates at the visual map tile zoom level
     // to pixel (or tile) coordinates at the opportunity density grid's zoom level, i.e. the number of visual map
     // pixels (or tiles) an opportunity grid pixel (or tile) is wide.
     // This is always a power of two, so as an optimization we could just store the difference in zoom levels
     // and scale using bitshift operators.
-    const pixelWidthAtZoom = Math.pow(2, zoom - this.grid.zoom)
-
-    // Find the top leftmost opportunity grid pixel falling inside the visual map tile we're rendering.
-    const gxMin = x / pixelWidthAtZoom - this.grid.west
-    const gyMin = y / pixelWidthAtZoom - this.grid.north
-
+    const zoomDifference = zoom - this.grid.zoom
+    // FIXME what happens below when zoomDifference is negative, i.e. when we're really zoomed out?
+    const gridCellWidthInTilePixels = Math.pow(2, zoomDifference)
+    const tileWidthInGridCells = 256 / gridCellWidthInTilePixels 
+    // The total number of tile pixels falling within one grid cell, i.e. the square of the width in pixels.
+    // This is used to convert job counts per cell to job counts per pixel.
+    // NOTE that this is instead proportional to the number of pixels in one dimension, which seems to scale better
+    // const tilePixelsPerGridCell = gridCellWidthInTilePixels * gridCellWidthInTilePixels * 150
+    // const tilePixelsPerGridCell = ZSCALE_BY_ZOOM[zoom] * 500
+    const tilePixelsPerGridCell = gridCellWidthInTilePixels * this.grid.max / 200
+    
+    // Find the top leftmost opportunity grid cell falling inside the map tile we're rendering.
+    const gxMin = mercPixelX / gridCellWidthInTilePixels - this.grid.west - 1
```

Comment:
Or I suppose it is not, given how you're doing the interpolation. Either way, a comment would be helpful.